# Chapter 2.3. - Qbin Operators and Operations in Expressions and Assignments

In this chapter we will be introducing *Qbin* operators and operations and talking about their differences. Also, we will use them to form more complex expressions and assignments by chaining multiple operations and operators together. In some cases, the use of complex expressions and assignments is quite strait forward. However, as in some cases complex expressions and assignments might result in unexpected solutions, we will investigate some of the possible consequences of chaining certain operators and operations.

First, let import the ***Qbin* class from d5 module of dann5 quantum programming package** and then initialize three *Qbin* variables, *x*, *y* and *z*, as we have discussed in chapter 1.3.

In [1]:
from dann5.d5 import Qbin
x = Qbin(2, "x"); y = Qbin(1, "y"); z = Qbin(2, "z");
print(x, y, z)

x\2q:U\ y\1q:U\ z\2q:U\


Also, we need to activate the default solver, which will be used to solve quantum problems in this chapter.

In [2]:
from dann5.dwave import Solver
print(Solver.Active())

## Qbin comparison operators

*Qbin* comparison operators create a relationship between one input and one output *Qbin* variable. In a way they are allowing a programmer to define entanglement rules between two quantum binary variables. Thus, the values in a valid solution are enforced on variables to satisfy the specific operator, e.g.
> x == y, creates an equal relationship between *Qbin* variables, which enforces that x and y are always equal in a valid solution.

In [3]:
eqXpr = x == y
print(eqXpr)
print(eqXpr.toString(True).replace(";", ";\n"))

(x\2q:U\ == y\1q:U\)
x0\S\ == y0\S\;
 x1\S\ == y1\0\;
 


From d5vc above we see that *Qbin* variable *y* has been extended to match the size of *Qbin* variable *x*, but to preserve its solution capacity the added *Qbit* has value *0*. Also, we see that for *x* and *y* to be equal all their q-bits need to be equal, i.e. in d4vc we are using *Qbit* equal operator (*==*) between respectful q-bits.
   
- **Note**: q-bit's within *Qbin* variable are identified by its position within the binary sequence and the name of the quantum-binary variable, i.e. the first q-bit within *Qbin* variable defined as "x" in quantum space is identified as x1.

The solution set bellow, reflexts both, the defined quantum capacities of variables *x* and *y*, and defined quantum comparison rule that their values have to be equal in a valid solution.

In [4]:
print(eqXpr.solve())

x\2b:00\; y\1b:0\
x\2b:01\; y\1b:1\



Similarly,
> *x != y*, creates a not equal relationship between *Qbin* variables, which enforces that *x* and *y* are always to have different values in a valid solution.

In [5]:
neXpr = x != z
print(neXpr)
print(neXpr.toString(True).replace(";", ";\n"))

(x\3q:U\ != z\3q:U\)
{ x0\S\ = _!=00\S\ .+ z0\S\;
 _1\1\ = _!=01\S\ | _!=00\S\;
 };
 { x1\S\ = _!=01\S\ + z1\S\ + #[x0]\S\;
 };
 { x2\S\ = #[x1]\S\;
 };
 


However, from d5vc we see that the implementation of the *Qbin* comparison not-equal (*!=*) operator requires a different approach than just simply using, in this case, *Qbit* not-equal operators. For two *Qbin* variables, *x* and *y* to be different it means that we need to define a rule where
> at least one q-bit of those 2 variables is different.

To achieve this we are comparing these *Qbin* variables using *Qwhole* not-equal (*!=*) operator described in the next chapter. As a result, we have the solution set as one in the cell below, where *Qbin* variables *x* and *z* have different binary sequences, which differ in at least one of their q-bits.

In [6]:
print(neXpr.solve())

x\2b:00\ z\2b:10\ 
x\2b:10\ z\2b:00\ 
x\2b:01\ z\2b:11\ 
x\2b:11\ z\2b:01\ 
x\2b:01\ z\2b:00\ 
x\2b:11\ z\2b:10\ 
x\2b:01\ z\2b:10\ 
x\2b:11\ z\2b:00\ 
x\2b:00\ z\2b:11\ 
x\2b:10\ z\2b:01\ 
x\2b:00\ z\2b:01\ 
x\2b:10\ z\2b:11\ 



- **Note**: Loke *Qbool*, there are only **equal (==)** and **not-equal (!=)** quantum comparison operators for *Qbin* variables.

## Qbin bitwise operator

Additionally, **inversion ~** (tilde) is a unary *Qbin* bitwise operator, where an automatically generated *Qbin* inverted variable is introduced. 
> In the following example an inversion of a *Qbin* variable with id *x* is a dann5 system generated *Qbin* variable *~x*.

From the first printout we see the inversion expression in quantum space. From the second printout we see valid solutions where each q-bit of *Qbin* variable *~x* has inverted value from the corresponding q-bit of original *Qbin* variable *x*.

In [7]:
xI = ~x
print(xI)
print(xI.solve())

(~x\2q:U\ ~ x\2q:U\)
~x\2b:11\; x\2b:00\
~x\2b:01\; x\2b:10\
~x\2b:10\; x\2b:01\
~x\2b:00\; x\2b:11\



## Qbin bitwise operations

*Qbin* bitwise operations always have two *Qbin* input arguments and at least a resulting *Qbin* output. Qbin **and(&)** and **or(|)** bitwise operations are those with one resulting *Qbin* variable. As we see in the example below, the resulting output variables are automatically generated with the names *_&(plus-number)* or *_|(plus-number)*.

In [8]:
andXpr = (x & y)
print(andXpr)
print(andXpr.toString(True))
print(andXpr.solve())

(x\2q:U\ & y\1q:U\)
_&00\S\ = x0\S\ & y0\S\; _&01\0\ = x1\S\ & y1\0\; 
_&0\2b:00\; x\2b:00\; y\1b:0\
_&0\2b:00\; x\2b:00\; y\1b:1\
_&0\2b:00\; x\2b:10\; y\1b:0\
_&0\2b:00\; x\2b:10\; y\1b:1\
_&0\2b:00\; x\2b:01\; y\1b:0\
_&0\2b:00\; x\2b:11\; y\1b:0\
_&0\2b:01\; x\2b:01\; y\1b:1\
_&0\2b:01\; x\2b:11\; y\1b:1\



The appropriate *Qbit* bitwise operations are applied on corresponding q-bits of two *Qbin* variables, as we see from the resulting solution sets, both for AND (*&*, above) and OR (*|*, below) operations.

In [9]:
orXpr = (x | y)
print(orXpr)
print(orXpr.toString(True))
print(orXpr.solve())

(x\2q:U\ | y\1q:U\)
_|10\S\ = x0\S\ | y0\S\; _|11\S\ = x1\S\ | y1\0\; 
_|1\2b:00\; x\2b:00\; y\1b:0\
_|1\2b:10\; x\2b:10\; y\1b:0\
_|1\2b:01\; x\2b:00\; y\1b:1\
_|1\2b:01\; x\2b:01\; y\1b:0\
_|1\2b:01\; x\2b:01\; y\1b:1\
_|1\2b:11\; x\2b:10\; y\1b:1\
_|1\2b:11\; x\2b:11\; y\1b:0\
_|1\2b:11\; x\2b:11\; y\1b:1\



*Qbin* **nand** and **nor** are quantum bitwise operations with two input arguments, which can be either *Qbin* variables or expressions (QbinExpr) and with dann5 system generated resulting and auxiliary *Qbin* variable outputs. These operations are implemented as *Qbin* Python methods and in quantum space represented with *!&* and *!|* symbols, respectfully. Thus, the resulting *Qbin* outputs, in the example below, are dann5 system generated names *_!&(plus-number)* or *_!|(plus-number)* and they can be seen in the solution. 

- **Note**: The auxiliary *Qbin* output variables with dann5 system generated names *?!&(plus-number)* or *?!|(plus-number)* will show up only in the corresponding QUBOs, as they are required nodes for the QUBO transformation, and they will be discussed in the QUBO chapter(s) of the section 4, which will cover Quantum Program Computing using QUBO.

In [10]:
nandXpr = x.nand(y)
print(nandXpr)
print(nandXpr.toString(True))
print(nandXpr.solve())

(x\2q:U\ !& y\1q:U\)
_!&00\S\ = x0\S\ !& y0\S\; _!&01\S\ = x1\S\ !& y1\0\; 
_!&0\2b:11\; x\2b:00\; y\1b:0\
_!&0\2b:11\; x\2b:00\; y\1b:1\
_!&0\2b:11\; x\2b:10\; y\1b:0\
_!&0\2b:11\; x\2b:10\; y\1b:1\
_!&0\2b:11\; x\2b:01\; y\1b:0\
_!&0\2b:11\; x\2b:11\; y\1b:0\
_!&0\2b:10\; x\2b:01\; y\1b:1\
_!&0\2b:10\; x\2b:11\; y\1b:1\



Again, for these  bitwise operations we see that they are applied on corresponding q-bits of two *Qbin* variables, for both quantum bitwise operations, NAND (*!&*) above, and NOR (*!|*) below.

In [11]:
norXpr = x.nor(y)
print(norXpr)
print(norXpr.toString(True))
print(norXpr.solve())

(x\2q:U\ !| y\1q:U\)
_!|00\S\ = x0\S\ !| y0\S\; _!|01\S\ = x1\S\ !| y1\0\; 
_!|0\2b:00\; x\2b:10\; y\1b:1\
_!|0\2b:00\; x\2b:11\; y\1b:0\
_!|0\2b:10\; x\2b:00\; y\1b:1\
_!|0\2b:10\; x\2b:01\; y\1b:0\
_!|0\2b:01\; x\2b:10\; y\1b:0\
_!|0\2b:11\; x\2b:00\; y\1b:0\
_!|0\2b:00\; x\2b:11\; y\1b:1\
_!|0\2b:10\; x\2b:01\; y\1b:1\



The *Qbin* **xor(^)** is a quantum bitwise operation with two input arguments, which can be *Qbin* variables or expressions, and with dann5 system generated resulting and auxiliary *Qbin* output variables. The resulting *Qbin* output, in below example, is *_^#*, where *#* is an autogenerated number.

- **Note**: The auxiliary *Qbin* output variable, e.g. *#[_^0]* will show up only in QUBO, as it is required node of the QUBO transformation. The QUBO transformations of quantum variables will be discussed in chapter 4.1 when we discuss QUBO transformations of d5vc.

In [12]:
xorXpr = (x ^ y)
print(xorXpr)
print(xorXpr.toString(True))
print(xorXpr.solve())

(x\2q:U\ ^ y\1q:U\)
_^00\S\ = x0\S\ ^ y0\S\; _^01\S\ = x1\S\ ^ y1\0\; 
_^0\2b:00\; x\2b:00\; y\1b:0\
_^0\2b:10\; x\2b:10\; y\1b:0\
_^0\2b:01\; x\2b:00\; y\1b:1\
_^0\2b:01\; x\2b:01\; y\1b:0\
_^0\2b:11\; x\2b:10\; y\1b:1\
_^0\2b:11\; x\2b:11\; y\1b:0\
_^0\2b:00\; x\2b:01\; y\1b:1\
_^0\2b:10\; x\2b:11\; y\1b:1\



Similar to *nand* and *nor* quantum bitwise operations **nxor** is implemented as a *Qbin* method and in the quantum space is  represented with *!^* symbol. The resulting *Qbin* output, in below example, is *_!^(#)*, where *#* is an autogenerated number. 

Same as the *xor* quantum operation, the quantum *nxor* operation has a second autogenerated output *Qbin* variable with a dann5 system generated name #[_!^(plus-number)], where (plus-number) is an autogenerated number, like #[_!^0]. As stated earlier, we will be covering these auxiliary *Qbin* variables in chapter 4.1. when we talk about QUBO transformations.

In [13]:
nxorXpr = x.nxor(y)
print(nxorXpr)
print(nxorXpr.toString(True))
print(nxorXpr.solve())

(x\2q:U\ !^ y\1q:U\)
_!^00\S\ = x0\S\ !^ y0\S\; _!^01\S\ = x1\S\ !^ y1\0\; 
_!^0\2b:00\; x\2b:10\; y\1b:1\
_!^0\2b:00\; x\2b:11\; y\1b:0\
_!^0\2b:10\; x\2b:00\; y\1b:1\
_!^0\2b:10\; x\2b:01\; y\1b:0\
_!^0\2b:01\; x\2b:10\; y\1b:0\
_!^0\2b:11\; x\2b:00\; y\1b:0\
_!^0\2b:01\; x\2b:11\; y\1b:1\
_!^0\2b:11\; x\2b:01\; y\1b:1\



- **NOTE**: *Qbin* **xor** and **unlike** operations are interchangeable, while **nxor** and **alike** operations are interchangeable.

Both **unlike** and **like** operations are implemented as *Qbin* methods, as shown in the example below.

In [14]:
unlikeXpr = x.unlike(z)
print(unlikeXpr)
print(unlikeXpr.toString(True))
print(unlikeXpr.solve())

(x\2q:U\ ^ z\2q:U\)
_^10\S\ = x0\S\ ^ z0\S\; _^11\S\ = x1\S\ ^ z1\S\; 
_^1\2b:00\; x\2b:00\; z\2b:00\
_^1\2b:10\; x\2b:00\; z\2b:10\
_^1\2b:10\; x\2b:10\; z\2b:00\
_^1\2b:01\; x\2b:00\; z\2b:01\
_^1\2b:01\; x\2b:01\; z\2b:00\
_^1\2b:11\; x\2b:00\; z\2b:11\
_^1\2b:11\; x\2b:10\; z\2b:01\
_^1\2b:11\; x\2b:01\; z\2b:10\
_^1\2b:11\; x\2b:11\; z\2b:00\
_^1\2b:00\; x\2b:10\; z\2b:10\
_^1\2b:01\; x\2b:10\; z\2b:11\
_^1\2b:01\; x\2b:11\; z\2b:10\
_^1\2b:00\; x\2b:01\; z\2b:01\
_^1\2b:10\; x\2b:01\; z\2b:11\
_^1\2b:10\; x\2b:11\; z\2b:01\
_^1\2b:00\; x\2b:11\; z\2b:11\



From the solution sets of both quantum *unlike* (above) and *alike* (below) operation expressions we see that for some solutions the relationship is enforced on all corresponding q-bits of two *Qbin* variables by the resulting variable _^1 having all q-bits with value 1. For the *unlike* operation (above) these are:
> _^1\2b:11\; x\2b:00\; z\2b:11\
>
> _^1\2b:11\; x\2b:10\; z\2b:01\
>
> _^1\2b:11\; x\2b:01\; z\2b:10\
>
> _^1\2b:11\; x\2b:11\; z\2b:00\

Or, for the *alike* operation (below) those are:
> _!^1\2b:11\; x\2b:00\; z\2b:00\
>
> _!^1\2b:11\; x\2b:10\; z\2b:10\
>
> _!^1\2b:11\; x\2b:01\; z\2b:01\
>
> _!^1\2b:11\; x\2b:11\; z\2b:11\

Also, we see that opposite solutions for *Qbin* variables *x* and *z* are correct when the rule is enforced with the resulting variable _^1 having all q-bits with value 0. For example, for *unlike* operation these solutions are:
> _^1\2b:00\; x\2b:00\; z\2b:00\
>
> _^1\2b:00\; x\2b:10\; z\2b:10\
>
> _^1\2b:00\; x\2b:01\; z\2b:01\
>
> _^1\2b:00\; x\2b:11\; z\2b:11\

So, *x* and *z* have the same values as in case of *alike* operation when *_!^1* had all q-bits equal *1*. 

- **Note**: We will see here, and in the following chapters of this and section 3, how to use just quantum expression or just quantum assignments, or how to combine them into quantum (programming) blocks and routines, to identify a right set of solutions for a particular problem.

In [15]:
alikeXpr = x.alike(z)
print(alikeXpr)
print(alikeXpr.toString(True))
print(alikeXpr.solve())

(x\2q:U\ !^ z\2q:U\)
_!^10\S\ = x0\S\ !^ z0\S\; _!^11\S\ = x1\S\ !^ z1\S\; 
_!^1\2b:00\; x\2b:00\; z\2b:11\
_!^1\2b:00\; x\2b:10\; z\2b:01\
_!^1\2b:00\; x\2b:01\; z\2b:10\
_!^1\2b:00\; x\2b:11\; z\2b:00\
_!^1\2b:10\; x\2b:00\; z\2b:01\
_!^1\2b:10\; x\2b:01\; z\2b:00\
_!^1\2b:01\; x\2b:00\; z\2b:10\
_!^1\2b:01\; x\2b:10\; z\2b:00\
_!^1\2b:11\; x\2b:00\; z\2b:00\
_!^1\2b:10\; x\2b:10\; z\2b:11\
_!^1\2b:10\; x\2b:11\; z\2b:10\
_!^1\2b:11\; x\2b:10\; z\2b:10\
_!^1\2b:01\; x\2b:01\; z\2b:11\
_!^1\2b:01\; x\2b:11\; z\2b:01\
_!^1\2b:11\; x\2b:01\; z\2b:01\
_!^1\2b:11\; x\2b:11\; z\2b:11\



## Solving a Qbin expressions

To start we will initialize six quantum binary (*Qbin*) variables (*bn0-bn4* and *bnr*) to be used in following complex expressions and assignments.

- **Note**: We see in the code below that multiple types of number systems (octal, decimal, hexadecimal, etc.) can be passed to the *Qbin* constructor using *Bits* from *d5* module of *dann5* package, and all will be initialized as discretionary *Qbin* variables.

In [16]:
from dann5.d5 import Bits

bn0 = Qbin("bn0", Bits(0o03)) # you can pass octal numbers
bn1 = Qbin(3, "bn1")
bn2 = Qbin(4, "bn2")
bn3 = Qbin("bn3", Bits(0b110)) # 6 in binary format
bn4 = Qbin(2, "bn4")
bnr = Qbin("bnr", Bits(0x5)) # you can pass hexadecimal numbers

print(bn0)
print(bn1)
print(bn2)
print(bn3)
print(bn4)
print(bnr)

bn0\2q:11\
bn1\3q:U\
bn2\4q:U\
bn3\3q:110\
bn4\2q:U\
bnr\3q:101\


As with any other quantum type, we can now form a complex quantum expression by using quantum bitwise operations, like *AND (&)*, *OR (|)* and *XOR(^)*, and by using a quantum comparison operator, like *EQUAL (==)*. 

In [17]:
expression = ((bn0 & bn1) | ((bn2 ^ bn3) == bn4))
print("\n {} \n\n {}\n".format(expression.toString(), 
                               expression.toString(True).replace(";", ";\n")))


 ((bn0\2q:11\ & bn1\3q:U\) | ((bn2\4q:U\ ^ bn3\3q:110\) == bn4\2q:U\)) 

 _|20\S\ = _&10\S\ | _^20\S\;
 _&10\S\ = bn00\1\ & bn10\S\;
 _^20\S\ == bn40\S\;
 _^20\S\ = bn20\S\ ^ bn30\0\;
 _|21\S\ = _&11\S\ | _^21\S\;
 _&11\S\ = bn01\1\ & bn11\S\;
 _^21\S\ == bn41\S\;
 _^21\S\ = bn21\S\ ^ bn31\1\;
 _|22\S\ = _&12\0\ | _^22\S\;
 _&12\0\ = bn02\0\ & bn12\S\;
 _^22\S\ == bn42\0\;
 _^22\S\ = bn22\S\ ^ bn32\1\;
 _|23\S\ = &3\0\ | _^23\S\;
 _^23\S\ == bn43\0\;
 _^23\S\ = bn23\S\ ^ bn33\0\;
 



From the *expression*'s d5vc above, we see that some variables will be auto-resized by *dann5* to ensure a uniformity of operation and truthfulness of calculated solutions.

1. The variable *bn0/2q:11/* has been defined with 2 q-bits and deterministic value *0b11 (3)*. However, due to *bn0 & bn1* sub-expression and the fact that bn1/3q:U/ has 3 q-bits, *bn0* variable within the quantum expression is modified by introduction of its 3rd q-bit *bn02\0\* in *_&02\0\ = bn02\0\ & bn12\S\* line of expression's d5vc.
2. Due to *bn2\4q:U\* having four q-bits, *bn3\3q:110\* and *bn4\2q:U\* variables within the quantum expression had to be modified and extended to four q-bit variables
    - *bn33\0\* q-bit has been added due to *_^03\S\ = bn23\S\ ^ bn33\0\* line, and
    - *bn42\0\* and *bn43\0\* q-bits have been added to *_^02\S\ == bn42\0\* and *_^03\S\ == bn43\0\* lines

To find the solutions to the problem described by quantum expression called *expression*, we will execute *expression.solve()* method.

In [18]:
print("dann5 simulator solutions: \n{}\n".format(expression.solve()))

dann5 simulator solutions: 
_|2\4b:0010\; _&1\3b:000\; bn0\2b:11\; bn1\3b:000\; _^2\4b:0010\; bn2\4b:0100\; bn3\3b:110\; bn4\2b:10\
_|2\4b:0010\; _&1\3b:000\; bn0\2b:11\; bn1\3b:100\; _^2\4b:0010\; bn2\4b:0100\; bn3\3b:110\; bn4\2b:10\
_|2\4b:0011\; _&1\3b:000\; bn0\2b:11\; bn1\3b:000\; _^2\4b:0011\; bn2\4b:0101\; bn3\3b:110\; bn4\2b:11\
_|2\4b:0011\; _&1\3b:000\; bn0\2b:11\; bn1\3b:100\; _^2\4b:0011\; bn2\4b:0101\; bn3\3b:110\; bn4\2b:11\
_|2\4b:0010\; _&1\3b:010\; bn0\2b:11\; bn1\3b:010\; _^2\4b:0010\; bn2\4b:0100\; bn3\3b:110\; bn4\2b:10\
_|2\4b:0010\; _&1\3b:010\; bn0\2b:11\; bn1\3b:110\; _^2\4b:0010\; bn2\4b:0100\; bn3\3b:110\; bn4\2b:10\
_|2\4b:0011\; _&1\3b:010\; bn0\2b:11\; bn1\3b:010\; _^2\4b:0011\; bn2\4b:0101\; bn3\3b:110\; bn4\2b:11\
_|2\4b:0011\; _&1\3b:010\; bn0\2b:11\; bn1\3b:110\; _^2\4b:0011\; bn2\4b:0101\; bn3\3b:110\; bn4\2b:11\
_|2\4b:0011\; _&1\3b:001\; bn0\2b:11\; bn1\3b:001\; _^2\4b:0010\; bn2\4b:0100\; bn3\3b:110\; bn4\2b:10\
_|2\4b:0011\; _&1\3b:001\; bn0\2b:11

By reviewing the calculated solutions we can see that the 2nd q-bit of auto-generated quantum variable named *_&0* will always be *0* due to the fact that undefined 2nd q-bit of *bn0* quantum variable was automatically set to be *0*. Similarly, q-bits 2 and 3 of *_^0* are always 0, due to being *EQUAL (==)* to bn4, which was automatically resized by adding q-bits 2 and 3 with values *0*.

We can adjust the *Qbin* varaibles, by resizing and setting the specific values of variables q-bits, like in the code cell below.

In [19]:
bn0.resize(3)
bn4.resize(4)
print("After resize:\t\t\t{}; {}\n".format(bn0, bn4.toString(True)))
      
bn0[2] = 1
bn4[3] = 0
print("After setting q-bit values:\t{}; {}".format(bn0, bn4.toString(True)))

After resize:			bn0\3q:011\; bn4\4q:bn43\S\;bn42\S\;bn41\S\;bn40\S\;\

After setting q-bit values:	bn0\3q:111\; bn4\4q:bn43\0\;bn42\S\;bn41\S\;bn40\S\;\


## Solving a quantum assignment

Now we can limit the scope of the solutions by defining a quantum assignment. In the code below we will request all quantum expression solutions that result in a binary sequence 0b101.

- **Note**: To assign an expression to a resulting *Qbin* variable we can use ***variable.assign(expression)*** or ***variable._(expression)*** methods.

If we make a quantum assignment by using the same expression (with modified *bn0* and *bn4* variables) and using *bnr\3q:101\* as an assignee, we can see from assignment's d5vc in the following code cell output that the second q-bit of *bn0*, *bn02* is set to 1, and the second q-bit of *bn2* is in superposition state, *bn22\S\*. Also, the *bnr* variable is extended and additional *bnr3* Qbit with value 0 is added.

In [20]:
qAssign = bnr._(((bn0 & bn1) | ((bn2 ^ bn3) == bn4)))
print("\n {} \n\n {}\n".format(qAssign.toString(), 
                               qAssign.toString(True).replace(";", ";\n")))


 bnr\4q:0101\ = ((bn0\3q:111\ & bn1\3q:U\) | ((bn2\4q:U\ ^ bn3\3q:110\) == bn4\4q:U\)) 

 bnr0\1\ = _&20\S\ | _^30\S\;
 _&20\S\ = bn00\1\ & bn10\S\;
 _^30\S\ == bn40\S\;
 _^30\S\ = bn20\S\ ^ bn30\0\;
 bnr1\0\ = _&21\S\ | _^31\S\;
 _&21\S\ = bn01\1\ & bn11\S\;
 _^31\S\ == bn41\S\;
 _^31\S\ = bn21\S\ ^ bn31\1\;
 bnr2\1\ = _&22\S\ | _^32\S\;
 _&22\S\ = bn02\1\ & bn12\S\;
 _^32\S\ == bn42\S\;
 _^32\S\ = bn22\S\ ^ bn32\1\;
 bnr3\0\ = &3\0\ | _^33\S\;
 _^33\S\ == bn43\0\;
 _^33\S\ = bn23\S\ ^ bn33\0\;
 



- **Note**:  Without setting the 3rd q-bit of *bn0* variable to 1 or allowing for the 3rd q-bit of *bn4* to be in superposition state, the quantum assignment *qAssign* would be illogical and we would not be able to calculate a single correct solution. 

We are basing this conclusion on the fact that the quantum assignment's expression is same as previously (above) discussed quantum expression named *expression*. However, when we inspect the solutions of the *expression*, we see that there is not a single solution where the resulting auto-generated quantum *Qbin* variable *_|2* has a binary value *0b0101*. Considering that by forming an assignment the deterministic *Qbin* variable *bnr* as an assignee will bind to the *expression* by replacing the system-generated *_|2 Qbin* variable, its deterministic value has to be within the set of valid solutions for *_|2*.

In [21]:
print("dann5 simulator solutions: \n{}\n".format(qAssign.solve()))

dann5 simulator solutions: 
bnr\4b:0101\; _&2\3b:000\; bn0\3b:111\; bn1\3b:000\; _^3\4b:0101\; bn2\4b:0011\; bn3\3b:110\; bn4\4b:0101\
bnr\4b:0101\; _&2\3b:100\; bn0\3b:111\; bn1\3b:100\; _^3\4b:0101\; bn2\4b:0011\; bn3\3b:110\; bn4\4b:0101\
bnr\4b:0101\; _&2\3b:001\; bn0\3b:111\; bn1\3b:001\; _^3\4b:0100\; bn2\4b:0010\; bn3\3b:110\; bn4\4b:0100\
bnr\4b:0101\; _&2\3b:001\; bn0\3b:111\; bn1\3b:001\; _^3\4b:0101\; bn2\4b:0011\; bn3\3b:110\; bn4\4b:0101\
bnr\4b:0101\; _&2\3b:101\; bn0\3b:111\; bn1\3b:101\; _^3\4b:0100\; bn2\4b:0010\; bn3\3b:110\; bn4\4b:0100\
bnr\4b:0101\; _&2\3b:101\; bn0\3b:111\; bn1\3b:101\; _^3\4b:0101\; bn2\4b:0011\; bn3\3b:110\; bn4\4b:0101\
bnr\4b:0101\; _&2\3b:100\; bn0\3b:111\; bn1\3b:100\; _^3\4b:0001\; bn2\4b:0111\; bn3\3b:110\; bn4\4b:0001\
bnr\4b:0101\; _&2\3b:101\; bn0\3b:111\; bn1\3b:101\; _^3\4b:0000\; bn2\4b:0110\; bn3\3b:110\; bn4\4b:0000\
bnr\4b:0101\; _&2\3b:101\; bn0\3b:111\; bn1\3b:101\; _^3\4b:0001\; bn2\4b:0111\; bn3\3b:110\; bn4\4b:0001\




In **the next chapter, 2.4**, we will be covering the ***Qwhole* Operators and Operations in Expressions and Assignments**. In chapter 3.2., we will expand and look at how to use *Qbin* expressions and assignments to solve practical problems.